taichi/transforms/lower_ast.cpp

```diff
@@ -167,11 +167,50 @@ class LowerAST : public IRVisitor {
       auto end = stmt->end;
       begin->flatten(flattened);
       end->flatten(flattened);
-      auto &&new_for = std::make_unique<RangeForStmt>(
-          stmt->parent->lookup_var(stmt->loop_var_id[0]), begin->stmt,
-          end->stmt, std::move(stmt->body), stmt->vectorize, stmt->parallelize,
-          stmt->block_dim, stmt->strictly_serialized);
-      flattened.push_back(std::move(new_for));
+      static bool is_good_range_for = true; // TODO(archibate): detect if outer-most one
+      if (1 || is_good_range_for) { // #578
+        is_good_range_for = false;
+        auto &&new_for = std::make_unique<RangeForStmt>(
+            stmt->parent->lookup_var(stmt->loop_var_id[0]), begin->stmt,
+            end->stmt, std::move(stmt->body), stmt->vectorize, stmt->parallelize,
+            stmt->block_dim, stmt->strictly_serialized);
+        flattened.push_back(std::move(new_for));
+      } else {
+        // transform into a structure as
+        // while (1) { cond; if (no active) break; original body...}
+        //auto loop_var = Stmt::make<AllocaStmt>(DataType::i32);
+        auto loop_var = stmt->parent->lookup_var(stmt->loop_var_id[0]);
+        flattened.push_back(Stmt::make<LocalStoreStmt>(loop_var, begin->stmt));
+        auto loop_var_addr = LaneAttribute<LocalAddress>(
+            LocalAddress(loop_var->as<AllocaStmt>(), 0));
+        // auto loop_var_store = Stmt::make<LocalStoreStmt>(loop_var_addr, begin->stmt);
+        auto loop_var_load = Stmt::make<LocalLoadStmt>(loop_var_addr);
+        auto cond_stmt = Stmt::make<BinaryOpStmt>(BinaryOpType::cmp_lt,
```

Comment:
May be it wants me push this two into flattened? std::unique_ptr so annoying! Don't know if get() work after std::move.

Reply:
Yes, you need to store the instructions in the block.